webrender/src/device/gfx/device.rs

```diff
@@ -1152,23 +1281,85 @@ impl<B: hal::Backend> Device<B> {
+            uMode: self.program_mode_id,
+        };
+
+        let bound_locals = if self.locals.contains_key(&locals) {
+            *self.locals.get(&locals).unwrap()
+        } else {
+            assert!(self.desc_pool_locals.next());
+            let (descriptor_set, idx) = self.desc_pool_locals.descriptor_set();
+            self.locals_buffer.add(&self.device, &[locals], &mut self.heaps);
+            unsafe {
+                self.device.write_descriptor_sets(Some(hal::pso::DescriptorSetWrite {
+                    set: descriptor_set,
+                    binding: DESCRIPTOR_SET_LOCALS as _,
+                    array_offset: 0,
+                    descriptors: Some(hal::pso::Descriptor::Buffer(
+                        &self.locals_buffer.buffer().buffer,
+                        Some(0) .. None,
+                    )),
+                }));
+            }
+            self.locals.insert(locals, idx);
+            idx
+        };
+        self.bound_locals = bound_locals;
+    }
+
     pub fn bind_textures(&mut self) {
         debug_assert!(self.inside_frame);
         assert_ne!(self.bound_program, INVALID_PROGRAM_ID);
         let program = self
             .programs
             .get_mut(&self.bound_program)
             .expect("Program not found.");
-        let desc_set = self.descriptor_pools[self.next_id].get(&program.shader_kind);
-        for &(index, sampler_name) in SAMPLERS[0 .. 5].iter() {
-            program.bind_texture(
-                &self.device,
-                desc_set,
-                &self.images[&self.bound_textures[index]].core,
-                sampler_name,
-            );
+
+        let need_alloc = {
+            let bound_resources = DescriptorSetResources::new(
+                program.shader_kind.descriptor_group(),
+                [self.bound_textures[SAMPLERS[0].0],
+                    self.bound_textures[SAMPLERS[1].0],
+                    self.bound_textures[SAMPLERS[2].0],
+                    self.bound_textures[SAMPLERS[3].0],
+                    self.bound_textures[SAMPLERS[4].0]],
+                );
+
+            let (desc_set, need_alloc) = if self.descriptors_per_draw[self.next_id].contains_key(&bound_resources) {
+                let (pool_idx, desc_idx) = self.descriptors_per_draw[self.next_id][&bound_resources];
+                (self.descriptor_pools[self.next_id].pool_at_idx(&program.shader_kind, pool_idx).descriptor_set_at_idx(desc_idx), false)
+            } else {
+                let (desc_set, pool_idx, set_idx) = self.descriptor_pools[self.next_id].get(&program.shader_kind);
+                let value = (pool_idx, set_idx);
+                self.descriptors_per_draw[self.next_id].insert(bound_resources, value);
+                (desc_set, true)
+            };
+            self.bound_desc_set_resources = bound_resources;
+
+            for &(index, sampler_name) in SAMPLERS[0 .. 5].iter() {
+                program.bind_texture(
```

Comment:
why are we binding textures per program? This is not matching Vulkan/gfx binding model